### PR TITLE
Make bitmap index only support equal query

### PR DESCRIPTION
**Pull request by lee-lei**
## What changes were proposed in this pull request?
1. Filter out range intervals before building bitmap scanner
2. If no equal value interval, bypass bitmap scanner
3. Add one test case to verify the above assumption
## How was this patch tested?
mvn clean test package

### CODE REVIEW

**Comment by lee-lei** [comment]
Please help review this PR. Thanks for your time.

**Comment by lee-lei** [comment]
@adrian-wang @linhong-intel @yueawang Thanks.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala [hunk 1]
```diff
@@ -224,6 +230,11 @@ private[oap] object DummyIndexContext extends IndexContext(null) {
 private[oap] class FilterOptimizer(keySchema: StructType) {
   val order = GenerateOrdering.create(keySchema)
 
+  def isSingleValueInterval(interval: RangeInterval): Boolean = {
+    if (interval.start == interval.end && interval.startInclude && interval.endInclude) true
```
Comment (yueawang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: return interval.start == interval.end && interval.startInclude && interval.endInclude

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala [hunk 2]
```diff
@@ -187,17 +187,23 @@ private[oap] class IndexContext(meta: DataSourceMeta) extends Logging {
...
-        keySchema = new StructType().add(meta.schema(entries(lastIdx)))
-        scanner = BitMapScanner(bestIndexer)
         val attribute = meta.schema(entries(lastIdx)).name
         val filterOptimizer = unapply(attribute).get
-        scanner.intervalArray =
+        val sortedIntervalArray =
           intervalMap(attribute).sortWith(filterOptimizer.compareRangeInterval)
+        val singleValueIntervalArray =
+          sortedIntervalArray.filter(filterOptimizer.isSingleValueInterval)
+        if (singleValueIntervalArray.size > 0) {
```
Comment (yueawang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: use nonEmpty??

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala [hunk 3]
```diff
@@ -224,6 +230,11 @@ private[oap] object DummyIndexContext extends IndexContext(null) {
 private[oap] class FilterOptimizer(keySchema: StructType) {
   val order = GenerateOrdering.create(keySchema)
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: Will this be used?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala [hunk 4]
```diff
@@ -187,17 +187,23 @@ private[oap] class IndexContext(meta: DataSourceMeta) extends Logging {
...
+          scanner = BitMapScanner(bestIndexer)
+          logDebug("Bitmap index only supports equal query.")
+          scanner.intervalArray = singleValueIntervalArray
+        }
       case _ =>
     }
 
     logDebug("Index Scanner Intervals: " + scanner.intervalArray.mkString(", "))
-    scanner.withKeySchema(keySchema)
+    if (scanner != null && keySchema != null) scanner.withKeySchema(keySchema)
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: Is it possible for `scanner == null`? Is it supposed to set the keySchema while we create the scanner.

Comment (chenghao-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: https://github.com/Intel-bigdata/OAP/pull/395/files#diff-c1beda8555c50f88113fa60658645d47R205

Comment (chenghao-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: https://github.com/Intel-bigdata/OAP/pull/395/files#diff-c1beda8555c50f88113fa60658645d47L199

**Comment by chenghao-intel** [comment]
Can you please add a unit test to verify weather the BitMapIndex will be triggered even if we specified a Range search?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala (hunk 3 above)
Comment (chenghao-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: I think we need to refactor the `IndexContext` or `IndexScanner`, `ordering` or `intervalArray`, `pattern` probably not necessary to put into the base class.

@linhong-intel @adrian-wang

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala (hunk 4 above)
Comment (lee-lei) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: This is common for all the index types. Therefore it's better to be after building the specific index type as the above. Thanks.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala [hunk 5]
```diff
@@ -187,17 +187,23 @@ private[oap] class IndexContext(meta: DataSourceMeta) extends Logging {
...
       case BitMapIndex(entries) =>
-        keySchema = new StructType().add(meta.schema(entries(lastIdx)))
-        scanner = BitMapScanner(bestIndexer)
         val attribute = meta.schema(entries(lastIdx)).name
         val filterOptimizer = unapply(attribute).get
-        scanner.intervalArray =
+        val sortedIntervalArray =
           intervalMap(attribute).sortWith(filterOptimizer.compareRangeInterval)
+        val singleValueIntervalArray =
+          sortedIntervalArray.filter(filterOptimizer.isSingleValueInterval)
```
[the remark this answers is not in the document]
Comment (lee-lei) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: This is the entry point to build the specific OAP index types. Before here, interval maps is already built, combined and optimized. I agreee with you to assert all the filtered intervals to be single value intervals before building Bitmap Scanner. Thanks.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala [hunk 6]
```diff
@@ -224,6 +236,10 @@ private[oap] object DummyIndexContext extends IndexContext(null) {
 private[oap] class FilterOptimizer(keySchema: StructType) {
   val order = GenerateOrdering.create(keySchema)
 
+  def isSingleValueInterval(interval: RangeInterval): Boolean = {
+    return interval.start == interval.end && interval.startInclude && interval.endInclude
```
Comment (yueawang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: return keyword is unnecessary.
def isSingleValueInterval(interval: RangeInterval) =
    interval.start == interval.end && interval.startInclude && interval.endInclude

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala (hunk 3 above)
Comment (yueawang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: From original bitmap concept's point of view, maybe hash is better than ordering, especially now you support only equal query.
You may consider it and refine in future.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala (hunk 6 above)
Comment (lee-lei) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: It's more concise. As the public method, it needs explicit return type. So we can just remove the return keyword and keep Boolean keyword. Thanks.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala (hunk 3 above)
Comment (lee-lei) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: Thanks.

**Comment by lee-lei** [comment]
@adrian-wang @linhong-intel Please help review and merge this PR if it's ready. Thanks.

**Pull request merged**